_posts/2020-07-29-parallelized-sampling/parallelized-sampling.Rmd

```diff
@@ -0,0 +1,290 @@
+
+In the following sections, we will explain and illustrate what `SWoR` means probability-wise, briefly outline some alternative solutions we have considered but were not completely satisfied with, and then deep-dive into exponential variates, a simple mathematical construct that made the ideal solution for this problem possible.
+
+If you cannot wait to jump into action, there is also a [section](#examples) in which we showcase example usages of `sdf_weighted_sample()` in `sparklyr`. In addition, you can go to this [pull request](https://github.com/sparklyr/sparklyr/pull/2606) to examine all implementation details of `sparklyr::sdf_weighted_sample()`.
+
+## How it all started
+
+Our journey started from a [Github issue](https://github.com/sparklyr/sparklyr/issues/2592) inquiring about the possibility of supporting the equivalent of `dplyr::sample_frac(..., weight = <weight_column>)` for Spark dataframes in `sparklyr`. For example,
+
+```{r}
+dplyr::sample_frac(mtcars, 0.25, weight = gear, replace = FALSE)
+```
+
+```
+##                    mpg cyl  disp  hp drat    wt  qsec vs am gear carb
+## Merc 280C         17.8   6 167.6 123 3.92 3.440 18.90  1  0    4    4
+## Chrysler Imperial 14.7   8 440.0 230 3.23 5.345 17.42  0  0    3    4
+## Fiat X1-9         27.3   4  79.0  66 4.08 1.935 18.90  1  1    4    1
+## Hornet Sportabout 18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2
+## Valiant           18.1   6 225.0 105 2.76 3.460 20.22  1  0    3    1
+## Porsche 914-2     26.0   4 120.3  91 4.43 2.140 16.70  0  1    5    2
+## Maserati Bora     15.0   8 301.0 335 3.54 3.570 14.60  0  1    5    8
+## Ferrari Dino      19.7   6 145.0 175 3.62 2.770 15.50  0  1    5    6
+```
+
+will randomly select one-fourth of all rows from a R dataframe named "mtcars" without replacement, using `mtcars$gear` as weights. We were unable to find any function implementing the weighted versions of `dplyr::sample_frac` among [Spark SQL built-in functions](https://spark.apache.org/docs/3.0.0/api/sql/index.html) in Spark 3.0 or in earlier versions, which means a future version of `sparklyr` will need to run its own weighted sampling algorithm to support such use cases.
+
+## What exactly is `SWoR` {#swor}
+
+The purpose of this section is to mathematically describe the probability distribution generated by `SWoR` in terms of $w_1, \dotsc, w_N$, so that readers can clearly see that the exponential-variate based algorithm presented in a subsequent section in fact samples from precisely the same probability distribution. Readers already having a crystal-clear mental picture of what `SWoR` entails should probably skip most of this section. The key take-away here is given $N$ rows $r_1, \dotsc, r_N$ and their weights $w_1, \dotsc, w_N$ and a desired sample size $n$, the probability of `SWoR` selecting $(r_1, \dotsc, r_n)$ is $\prod\limits_{j = 1}^{n} \left( {w_j} \middle/ {\sum\limits_{k = j}^{N}{w_k}} \right)$.
```

Comment:
I think you can address the "Readers" as "you" here to give it a more conversational tone. ~ _"If you already have a clear mental picture of what `SWoR` entails, you can skip most of this section."_

note: I took out "crystal" from "crystal clear" because it _verges_ on sounding sarcastic (I know that's not the intent at all, but since there's no real value add to keeping it in, might as well tread lightly).